pkg/vm/json_test.go

```diff
@@ -0,0 +1,406 @@
+		EStack             []vmUTStackItem `json:"evaluationStack"`
+	}
+)
+
+const (
+	vmExecute  vmUTActionType = "Execute"
+	vmStepInto vmUTActionType = "StepInto"
+	vmStepOut  vmUTActionType = "StepOut"
+	vmStepOver vmUTActionType = "StepOver"
+
+	typeBoolean   vmUTStackItemType = "Boolean"
+	typeByteArray vmUTStackItemType = "ByteArray"
+	typeString    vmUTStackItemType = "String"
+	typeInterop   vmUTStackItemType = "Interop"
+	typeInteger   vmUTStackItemType = "Integer"
+	typeArray     vmUTStackItemType = "Array"
+	typeStruct    vmUTStackItemType = "Struct"
+	typeMap       vmUTStackItemType = "Map"
+
+	testsDir = "neo-vm/tests/neo-vm.Tests/Tests/"
+)
+
+func TestUT(t *testing.T) {
+	err := filepath.Walk(testsDir, func(path string, info os.FileInfo, err error) error {
+		if !strings.HasSuffix(path, ".json") {
+			return nil
+		}
+
+		testFile(t, path)
+		return nil
+	})
+
+	require.NoError(t, err)
+}
+
+func testFile(t *testing.T, filename string) {
+	data, err := ioutil.ReadFile(filename)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// FIXME remove when updating to NEO 3.0 where BOM is removed
+	if len(data) > 2 && data[0] == 0xef && data[1] == 0xbb && data[2] == 0xbf {
+		data = data[3:]
+	}
+
+	ut := new(vmUT)
+	if err = json.Unmarshal(data, ut); err != nil {
+		t.Fatal(err)
+	}
+
+	t.Run(ut.Category+":"+ut.Name, func(t *testing.T) {
+		for i := range ut.Tests {
+			test := ut.Tests[i]
+			t.Run(ut.Tests[i].Name, func(t *testing.T) {
+				prog := []byte(test.Script)
+				vm := load(prog)
+				vm.state = breakState
+
+				// FIXME remove in NEO 3.0
+				vm.getScript = getScript(test.ScriptTable)
+
+				// FIXME in NEO 3.0 tests it has id []byte{0x77, 0x77, 0x77, 0x77}
+				vm.RegisterInteropFunc("Test.ExecutionEngine.GetScriptContainer", InteropFunc(func(v *VM) error {
+					v.estack.Push(&Element{value: (*InteropItem)(nil)})
+					return nil
+				}), 0)
+				vm.RegisterInteropFunc("System.ExecutionEngine.GetScriptContainer", InteropFunc(func(v *VM) error {
+					v.estack.Push(&Element{value: (*InteropItem)(nil)})
+					return nil
+				}), 0)
+
+				for i := range test.Steps {
+					execStep(t, vm, test.Steps[i])
+					result := test.Steps[i].Result
+					require.Equal(t, State(result.State), vm.state)
+					if result.State == vmUTState(faultState) { // do not compare stacks on fault
+						continue
+					}
+
+					if len(result.InvocationStack) > 0 {
+						for i, s := range result.InvocationStack {
+							ctx := vm.istack.Peek(i).Value().(*Context)
+							if ctx.nextip < len(ctx.prog) {
+								require.Equal(t, s.InstructionPointer, ctx.nextip)
+								require.Equal(t, s.Instruction, Instruction(ctx.prog[ctx.nextip]).String())
+							}
+							compareStacks(t, s.EStack, vm.estack)
+							compareStacks(t, s.AStack, vm.astack)
+						}
+					}
+
+					if len(result.ResultStack) != 0 {
+						compareStacks(t, result.ResultStack, vm.estack)
+					}
+				}
+			})
+		}
+	})
+}
+
+func hash160(buf []byte) (h util.Uint160) {
```

Comment:
we have hash160 in `crypto/hash` package. So it's good to reuse or even add documentation why you're not doing it.

Reply:
Yeah, didn't see it.